api/src/test/java/ai/djl/ndarray/NDScopeTest.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.ndarray;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class NDScopeTest {
+
+    @Test
+    @SuppressWarnings("try")
+    public void testNDScope() {
+        NDArray detached;
+        NDArray inside;
+        NDArray uninvolved;
+        try (NDManager manager = NDManager.newBaseManager()) {
+            try (NDScope scope = new NDScope()) {
+                scope.suppressNotUsedWarning();
+                try (NDScope ignore = new NDScope()) {
+                    uninvolved = manager.create(new int[] {1});
+                    uninvolved.close();
+                    inside = manager.create(new int[] {1});
+                    // not tracked by any NDScope, but still managed by NDManager
+                    NDScope.unregister(inside);
+                }
+
+                detached = manager.create(new int[] {1});
+                detached.detach(); // detached from NDManager
```

Comment:
1. In most cases, when `NDArray.detach()` is called, user want to return the NDArray from the function.
2. And usually, NDManager is at the out side if NDScope

Which means user has to call both function in most of time. And we documented that once NDArray is detached, user must manually close it.

So I feel we should unregister from the NDScope when `NDArray.detach()` is called.

Reply:
Looks like we are abusing `NDArray.detach()` in many internal place (`attache()`, they should only detach from NDManager only. Introducing two `detach(boolean)` function seems too much. Requires user to call both should be fine for now.

Reply:
What is your intention to introduce NDScope?

Reply:
My intention would be: Implement it because you need it to close a memory leak that you cannot solve otherwise. So I would not use NDScope if I did not have to. But if I have a leak - which I unfortunately do - I would put the NDScope where I think no NDArray should leak through - no matter if inside or outside of NDManager. And in that case I would not want it to leak through for whatever reason.

Reply:
> Looks like we are abusing `NDArray.detach()` in many internal place (`attache()`, they should only detach from NDManager only. Introducing two `detach(boolean)` function seems too much. Requires user to call both should be fine for now.

I agree that if there were a convenience method where the user exactly knows that she/he also unregister from the NDScope there should be no problem.

Reply:
Should we leave as it is for now?